src/solc_0.7/GameToken/GameToken.sol

```diff
@@ -0,0 +1,436 @@
+        _asset = asset;
+        _nextId = 1;
+    }
+
+    ///////////////////////////////  Modifiers //////////////////////////////
+
+    modifier minterGuard() {
+        require(msg.sender == _minter || _minter == address(0), "INVALID_MINTER");
+        _;
+    }
+
+    modifier onlyOwnerOrEditor(uint256 id) {
+        require(msg.sender == _ownerOf(id) || _gameEditors[id][msg.sender], "OWNER_EDITOR_ACCESS_DENIED");
+        _;
+    }
+
+    modifier notToZero(address to) {
+        require(to != address(0), "DESTINATION_ZERO_ADDRESS");
+        _;
+    }
+
+    ///////////////////////////////  Functions //////////////////////////////
+
+    /// @notice Function to remove multiple assets from a GAME
+    /// @param gameId The GAME to remove assets from
+    /// @param assetId The asset Id to remove
+    /// @param to The address to send the removed asset to
+    /// @param uri The URI string to update the GAME token's URI
+    function removeSingleAsset(
+        uint256 gameId,
+        uint256 assetId,
+        address to,
+        string memory uri
+    ) external override minterGuard() onlyOwnerOrEditor(gameId) notToZero(to) {
+        _gameData[gameId]._values[assetId] = _gameData[gameId]._values[assetId].sub(1);
+        uint256 remainingAssets = _gameData[gameId]._values[assetId];
+
+        if (remainingAssets == 0) {
+            // "remove" is from EnumerableSet.sol
+            require(_gameData[gameId]._assets.remove(assetId), "ASSET_NOT_IN_GAME");
+        }
+
+        _asset.safeTransferFrom(address(this), to, assetId, 1, "");
+        uint256[] memory assets = new uint256[](1);
+        uint256[] memory values = new uint256[](1);
+        assets[0] = assetId;
+        values[0] = uint256(1);
+
+        setTokenURI(gameId, uri);
+        emit AssetsRemoved(gameId, assets, values, to);
+    }
+
+    /// @notice Function to remove multiple assets from a GAME
+    /// @param gameId The GAME to remove assets from
+    /// @param assetIds An array of asset Ids to remove
+    /// @param values An array of the number of each asset id to remove
+    /// @param to The address to send removed assets to
+    /// @param uri The URI string to update the GAME token's URI
+    function removeMultipleAssets(
+        uint256 gameId,
+        uint256[] memory assetIds,
+        uint256[] memory values,
+        address to,
+        string memory uri
+    ) public override minterGuard() onlyOwnerOrEditor(gameId) notToZero(to) {
+        (uint256 assetTypes, ) = getNumberOfAssets(gameId);
+        require(assetIds.length == values.length && assetIds.length <= assetTypes, "INVALID_INPUT_LENGTHS");
+        for (uint256 i = 0; i < assetIds.length; i++) {
+            // "remove" is from EnumerableSet.sol
+            uint256 assetValues = _gameData[gameId]._values[assetIds[i]];
+            if (values[i] >= _gameData[gameId]._values[assetIds[i]]) {
+                _gameData[gameId]._assets.remove(assetIds[i]);
```

Comment:
I suppose this is unnecessarily gas inefficient, as I suppose the enumerable set will do extra storage since it need to remove one by one
it might be better to implement a batch remove for that

Reply:
@wighawag Can you elaborate on " implement a batch remove" here?
As we're switching to a simpler data model (just using a mapping `mapping(uint256 => mapping(uint256 => uint256)) private _gameAssets;`  instead of an enumerableSet, maybe this comment is no longer relevant?
I'll be using `safeBatchTransferFrom` fro the actual transfer.

Reply:
calling this resolved as we no longer use enumerableSet.